From dab8c562ee24726f988f1acf8f9dfad80610a0be Mon Sep 17 00:00:00 2001
From: Liu Ying <Ying.Liu@freescale.com>
Date: Thu, 21 Jul 2011 16:47:04 +0800
Subject: [PATCH 150/256] ENGR00153462 IPUv3 lib:Improve the logic of fitting
 fb

This patch improves the logic of fitting framebuffer
for show-to-fb case. The desired output resolution
should be aligned with the original one, and the
desired virtual yres should be less than or equal to
the original one, otherwise, we reconfigure
the framebuffer.

Signed-off-by: Liu Ying <Ying.Liu@freescale.com>
---
 ipu/mxc_ipu_hl_lib.c | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/ipu/mxc_ipu_hl_lib.c b/ipu/mxc_ipu_hl_lib.c
index 65931a0..6635c11 100644
--- a/ipu/mxc_ipu_hl_lib.c
+++ b/ipu/mxc_ipu_hl_lib.c
@@ -980,11 +980,11 @@ done:
 static int fit_fb_setting(struct fb_var_screeninfo * var, int width,
 	int height, int fmt, ipu_channel_t fb_chan, int bufs)
 {
-	if (var->yoffset != 0)
-		return 0;
 	if (fb_chan == MEM_BG_SYNC)
 		return ((var->xres_virtual == var->xres) &&
-			(var->yres_virtual == bufs*var->yres));
+			(var->xres == (unsigned int)width) &&
+			(var->yres == (unsigned int)height) &&
+			(var->yres_virtual >= bufs*var->yres));
 
 	if ((colorspaceofpixel(fmt) == YUV_CS) &&
 			(var->nonstd != (unsigned int)fmt))
@@ -995,14 +995,15 @@ static int fit_fb_setting(struct fb_var_screeninfo * var, int width,
 		return 0;
 	if (fb_chan == MEM_DC_SYNC)
 		return ((var->xres_virtual == var->xres) &&
-			(var->yres_virtual == bufs*var->yres));
-	if (fb_chan == MEM_FG_SYNC) {
+			(var->xres == (unsigned int)width) &&
+			(var->yres == (unsigned int)height) &&
+			(var->yres_virtual >= bufs*var->yres));
+	if (fb_chan == MEM_FG_SYNC)
 		return ((var->xres == (unsigned int)width) &&
 			(var->xres_virtual == (unsigned int)width) &&
 			(var->yres == (unsigned int)height) &&
-			(var->yres_virtual == (unsigned int)(bufs*height)) &&
+			(var->yres_virtual >= (unsigned int)(bufs*height)) &&
 			(var->bits_per_pixel == fmt_to_bpp(fmt)));
-	}
 
 	return 1;
 }
-- 
1.8.0

